### ISSUE TITLE
Disconnect (index out of range when reading "ENTITY_POSITION")

### ISSUE THREAD

**Issue by Kaillera** [Regalis11/Barotrauma#3927]
- [x] I have searched the issue tracker to check if the issue has already been reported.

**Description**
When you're playing multiplayer with more than just the host. Shooting Piercing Ammo on crawler/mudraptor swarms that are next to the sub (probably latched) using a drone have a chance to error all the players except the host out to Entity_Position.

**Version**
0.10.5.1

**Additional Informations**
Not sure if other ammo will cause the same issue. This has happened multiple times.

[crashreport_object.log](https://github.com/Regalis11/Barotrauma/files/5291704/crashreport_object.log)
[event_error_log_client_Kaillera_629 PM.log](https://github.com/Regalis11/Barotrauma/files/5291702/event_error_log_client_Kaillera_629.PM.log)
[event_error_log_client_Kaillera_628 PM.log](https://github.com/Regalis11/Barotrauma/files/5291703/event_error_log_client_Kaillera_628.PM.log)
[event_error_log_server_Kaillera_1829.log](https://github.com/Regalis11/Barotrauma/files/5294107/event_error_log_server_Kaillera_1829.log)
[event_error_log_server_Kaillera_1828.log](https://github.com/Regalis11/Barotrauma/files/5294108/event_error_log_server_Kaillera_1828.log)
[event_error_log_server_D.B. Cooper_1829.log](https://github.com/Regalis11/Barotrauma/files/5294111/event_error_log_server_D.B.Cooper_1829.log)
[event_error_log_server_DarkRider_1829.log](https://github.com/Regalis11/Barotrauma/files/5294112/event_error_log_server_DarkRider_1829.log)
[event_error_log_server_Ghostface_1828.log](https://github.com/Regalis11/Barotrauma/files/5294113/event_error_log_server_Ghostface_1828.log)
[event_error_log_server_D.B. Cooper_1828.log](https://github.com/Regalis11/Barotrauma/files/5294114/event_error_log_server_D.B.Cooper_1828.log)

**Comment by Regalis11** [comment]
What's an "Entity_Position" error? Do you mean something like the errors reported in https://github.com/Regalis11/Barotrauma/issues/3019 ("index out of range", with the previous header being ENTITY_POSITION)?

**Comment by Kaillera** [comment]
Forgot to put caused by drone in title and added the error to additional informations.

**Comment by Regalis11** [comment]
Discovered a potential cause for this: I noticed that based on the error reports it seems to happen when reading data for coilgun/railgun ammo and characters (usually monsters that appear in large numbers) in succession. My theory is that when the projectiles and enemies are being rapidly created and removed, the clients may not have received all the EntitySpawner events when reading the position data (i.e. they may not be up-to-date on which items/characters currently exist), and can get the entities mixed up. If they don't know that e.g. a hammerhead spawn's ID has been taken up by a newly spawned coilgun bolt, they'll try to read the bolt's data with Character.ClientRead, which will end up going out of bounds since.

Pushed a fix in https://github.com/Regalis11/Barotrauma-development/commit/930baada8780da20be6212a80dd6be3183ecdbf3, let's see if it takes care of this.
